src/bscPlugin/codeActions/XmlFileCodeActionsProcessor.ts

```diff
@@ -0,0 +1,44 @@
+import type { CodeAction, Range } from 'vscode-languageserver';
+import { CodeActionKind } from 'vscode-languageserver';
+import { DiagnosticCodeMap } from '../../DiagnosticMessages';
+import type { XmlFile } from '../../files/XmlFile';
+import type { BsDiagnostic } from '../../interfaces';
+import util from '../../util';
+
+export class XmlFileCodeActionsProcessor {
+    public constructor(
+        public file: XmlFile,
+        public range: Range,
+        public diagnostics: BsDiagnostic[],
+        public codeActions: CodeAction[]
+    ) {
+
+    }
+
+    public process() {
+        for (const diagnostic of this.diagnostics) {
+            if (diagnostic.code === DiagnosticCodeMap.xmlComponentMissingExtendsAttribute) {
+                this.addMissingExtends();
+            }
+        }
+    }
+
+    private addMissingExtends() {
+        //add the attribute at the end of the first attribute, or after the `<component` if no attributes
+        const pos = (this.file.parser.ast.component.attributes[0] ?? this.file.parser.ast.component.tag).range.end;
```

Comment:
I'm not sure the position is correct for a tag without any attribute. You probably should just always insert at the end of the tag instead of trying to find the first attribute.

Also maybe have some `const { component } = this.file.parser.ast;` for readability.

Reply:
I did it this way so we don't occasionally insert two consecutive whitespaces. It's easier to reason about the whitespace after an attribute. Although I do agree that we should add this to the very end of the tag, so I adjusted this to find the FINAL attribute first, and then after the `<component` object second if there are no attributes

Reply:
Adding an attribute or a tag should be done with a helper function BTW, so it's testable and reusable.

Reply:
Good point. I'll do that.